### PR TITLE
[core] Fix HashBucketAssigner load index too large error with refactor exception

### PR DESCRIPTION
**Pull request by xuzifu666**
### Purpose
HashCommon in Int2ShortHashMap would cause IllegalArgumentException when expected > 1073741824L，need make regression to default constructor to continue in this condition.
relate to https://github.com/apache/paimon/issues/3776

### CODE REVIEW

File: paimon-common/src/main/java/org/apache/paimon/utils/Int2ShortHashMap.java [hunk 1]
```diff
@@ -67,7 +67,13 @@ public void put(int key, short value) {
         }
 
         public Int2ShortHashMap build() {
-            Int2ShortHashMap map = new Int2ShortHashMap(keyList.size());
+            Int2ShortHashMap map;
+            try {
+                map = new Int2ShortHashMap(keyList.size());
+            } catch (IllegalArgumentException e) {
+                map = new Int2ShortHashMap();
```
Comment (JingsongLi) [inline paimon-common/src/main/java/org/apache/paimon/utils/Int2ShortHashMap.java:74]: Is there a way to solve it without downgrading?

**Comment by xuzifu666** [comment]
@izhangzhihao  Can you try the pr first to comfirm it can be work?  Expect for your feedback ~

**Comment by izhangzhihao** [comment]
I will try this tomorrow

**Comment by xuzifu666** [comment]
> I will try this tomorrow

thanks~

**Comment by izhangzhihao** [comment]
Just created a simple unit test:

```java
public class HashBucketAssignerTest {

    @Test
    public void testFunction() {
        IntStream stream = IntStream.rangeClosed(1, 1_100_000_000);

        final Int2ShortHashMap.Builder builder = Int2ShortHashMap.builder();


        stream.forEach(
                i -> builder.put(i, (short)0)
        );

        builder.build();
    }
}
```

Error:

```log
java.lang.IllegalArgumentException: Too large (805306401 expected elements with load factor 0.75)

	at it.unimi.dsi.fastutil.HashCommon.arraySize(HashCommon.java:208)
	at it.unimi.dsi.fastutil.ints.Int2ShortOpenHashMap.insert(Int2ShortOpenHashMap.java:257)
	at it.unimi.dsi.fastutil.ints.Int2ShortOpenHashMap.put(Int2ShortOpenHashMap.java:265)
	at xx.Int2ShortHashMap.put(Int2ShortHashMap.java:20)
	at xx.Int2ShortHashMap$Builder.build(Int2ShortHashMap.java:62)
	at xx.HashBucketAssignerTest.testFunction(HashBucketAssignerTest.java:25)
```

**Comment by xuzifu666** [comment]
> Just created a simple unit test:
> 
> ```java
> public class HashBucketAssignerTest {
> 
>     @Test
>     public void testFunction() {
>         IntStream stream = IntStream.rangeClosed(1, 1_100_000_000);
> 
>         final Int2ShortHashMap.Builder builder = Int2ShortHashMap.builder();
> 
> 
>         stream.forEach(
>                 i -> builder.put(i, (short)0)
>         );
> 
>         builder.build();
>     }
> }
> ```
> 
> Error:
> 
> ```
> java.lang.IllegalArgumentException: Too large (805306401 expected elements with load factor 0.75)
> 
> 	at it.unimi.dsi.fastutil.HashCommon.arraySize(HashCommon.java:208)
> 	at it.unimi.dsi.fastutil.ints.Int2ShortOpenHashMap.insert(Int2ShortOpenHashMap.java:257)
> 	at it.unimi.dsi.fastutil.ints.Int2ShortOpenHashMap.put(Int2ShortOpenHashMap.java:265)
> 	at xx.Int2ShortHashMap.put(Int2ShortHashMap.java:20)
> 	at xx.Int2ShortHashMap$Builder.build(Int2ShortHashMap.java:62)
> 	at xx.HashBucketAssignerTest.testFunction(HashBucketAssignerTest.java:25)
> ```

Hi， do you had apply the patch，I had test the ut it in my local，not report the exception only oom error. @izhangzhihao

**Comment by izhangzhihao** [comment]
For your oom error, you need add `-Xmx32g`

**Comment by xuzifu666** [comment]
> For your oom error, you need add `-Xmx32g`

Yes，my compute memory not enough to so large，so you had tested in your flink job? @izhangzhihao

**Comment by izhangzhihao** [comment]
Can you find a 32G machine and run `java -jar` just for this test? Since it does not pass the unit test, there is no reason to believe it can pass a real flink job...

**Comment by xuzifu666** [comment]
OK，but from your error strack，I can not find the source position in HashCommon which is litter strange.

**Comment by xuzifu666** [comment]
> Can you find a 32G machine and run `java -jar` just for this test? Since it does not pass the unit test, there is no reason to believe it can pass a real flink job...

OK，but from your error strack，I can not find the source position in HashCommon which is litter strange.

**Comment by izhangzhihao** [comment]
Error happens here: 

```java
public static int arraySize(final int expected, final float f) {
		final long s = Math.max(2, nextPowerOfTwo((long)Math.ceil(expected / f)));
		if (s > (1 << 30)) throw new IllegalArgumentException("Too large (" + expected + " expected elements with load factor " + f + ")");
		return (int)s;
	}
```

**Comment by xuzifu666** [comment]
> arraySize

It seems not consistent with master paimon dependecies，would you change it the same with master？From the patch it should not run to the logic as your error stack during Int2ShortOpenHashMap build.

**Comment by izhangzhihao** [comment]
the master branch is using `8.5.12`, but I can give it a try(master will update to a newer version in the future anyway)

**Comment by xuzifu666** [comment]
> the master branch is using `8.5.12`, but I can give it a try(master will update to a newer version in the future anyway)

OK，apply the patch to the code，Thanks~ @izhangzhihao

**Comment by izhangzhihao** [comment]
exactly the same error:
```log
java.lang.IllegalArgumentException: Too large (805306401 expected elements with load factor 0.75)

	at it.unimi.dsi.fastutil.HashCommon.arraySize(HashCommon.java:208)
	at it.unimi.dsi.fastutil.ints.Int2ShortOpenHashMap.insert(Int2ShortOpenHashMap.java:257)
	at it.unimi.dsi.fastutil.ints.Int2ShortOpenHashMap.put(Int2ShortOpenHashMap.java:265)
```

And it took 10 mins:

image

**Comment by xuzifu666** [comment]
> exactly the same error:
> 
> ```
> java.lang.IllegalArgumentException: Too large (805306401 expected elements with load factor 0.75)
> 
> 	at it.unimi.dsi.fastutil.HashCommon.arraySize(HashCommon.java:208)
> 	at it.unimi.dsi.fastutil.ints.Int2ShortOpenHashMap.insert(Int2ShortOpenHashMap.java:257)
> 	at it.unimi.dsi.fastutil.ints.Int2ShortOpenHashMap.put(Int2ShortOpenHashMap.java:265)
> ```
> 
> And it took 10 mins:
> 
> image

OK， please show more error stack about paimon called

**Comment by xuzifu666** [comment]
OK，Int2ShortOpenHashMap insert operation need rehash array which would occur the error. Thanks for your support. @izhangzhihao

**Review by izhangzhihao — COMMENTED** [review]
So is this not a fix but an exception msg improvement?

**Comment by xuzifu666** [comment]
> So is this not a fix but an exception msg improvement?

Yes, currently we can give more advise to user about it. @izhangzhihao

File: paimon-common/src/main/java/org/apache/paimon/utils/Int2ShortHashMap.java [hunk 2]
```diff
@@ -67,10 +67,18 @@ public void put(int key, short value) {
...
-            for (int i = 0; i < keyList.size(); i++) {
-                map.put(keyList.getInt(i), valueList.getShort(i));
+            Int2ShortHashMap map;
+            try {
+                map = new Int2ShortHashMap(keyList.size());
+                for (int i = 0; i < keyList.size(); i++) {
+                    map.put(keyList.getInt(i), valueList.getShort(i));
+                }
+            } catch (IllegalArgumentException e) {
+                throw new PaimonUtilsException(
```
[the remark this answers is not in the document]
Comment (xuzifu666) [reply paimon-common/src/main/java/org/apache/paimon/utils/Int2ShortHashMap.java]: OK，had change exception to RumtimeException. @JingsongLi

**Comment by JingsongLi** [comment]
+1

**Review by izhangzhihao — APPROVED** [review]
The error message looks good to me; adding parallelism is a workaround. Ideally, the resource allocation for the task should match the data flow, rather than matching the total amount of data at rest. The current situation does not meet this ideal. Is there any plan for optimization?

**Pull request merged**